persistent/Database/Persist/Quasi/Internal.hs

```diff
@@ -928,28 +925,72 @@ getSqlNameOr def =
             _ ->
                 Nothing
 
+data EntityConstraintDefs = EntityConstraintDefs
+    { entityConstraintDefsIdField :: Maybe UnboundIdDef
+    , entityConstraintDefsPrimaryComposite :: Maybe UnboundCompositeDef
+    , entityConstraintDefsUniques :: Maybe (NonEmpty UniqueDef)
+    , entityConstraintDefsForeigns :: Maybe (NonEmpty UnboundForeignDef)
+    }
+
+instance Semigroup EntityConstraintDefs where
+    a <> b =
+        EntityConstraintDefs
+            { entityConstraintDefsIdField = just1 (entityConstraintDefsIdField a) (entityConstraintDefsIdField b)
+            , entityConstraintDefsPrimaryComposite = just1 (entityConstraintDefsPrimaryComposite a) (entityConstraintDefsPrimaryComposite b)
+            , entityConstraintDefsUniques = entityConstraintDefsUniques a <> entityConstraintDefsUniques b
+            , entityConstraintDefsForeigns = entityConstraintDefsForeigns a <> entityConstraintDefsForeigns b
+            }
+
+just1 :: (Show x) => Maybe x -> Maybe x -> Maybe x
+just1 (Just x) (Just y) = error $ "expected only one of: "
+  `mappend` show x `mappend` " " `mappend` show y
+just1 x y = x `mplus` y
```

Comment:
I went with using Semigroups/Monoids here but this implementation of `just1` which is used in the semigroup instance makes it unlawful. I could re-write this to not use these typeclasses, but I thought it was a nice refactoring (means I could use `foldMap`, `mempty`).

Alternatively this could use something like `First` or `Last`, though it is probably a nicer user experience to receive this error than for it to silently proceed using one of the values over the other in the case of duplicate values. In fact that is definitely not what we would want.

It may be that now I have done an initial refactor, I could tweak things further again to not have unlawful instances, and use something instead `foldMap` / `mempty`

Reply:
Hmm! I like using the `Monoid` and `Semigroup` instances here. That's neat. But I don't like having a `Semigroup` operation that will `error` on a *relatively* common case - that seems like a footgun.

I feel like a type that better represents this is something like:

```haskell
data SetOnce a 
    = SetOnce a 
     | NotSet 
     | SetManyTimes (NonEmpty a)

instance Semigroup (SetOnce a) where
    NotSet <> a = a

    SetOnce a <> NotSet = SetOnce a
    SetOnce a <> SetOnce b = SetManyTimes (a :| [b])
    SetOnce a <> SetManyTimes as = SetManyTimes (NEL.cons a as)

    a@(SetManyTimes xs) <> NotSet = a
    SetManyTimes xs <> SetOnce a = SetManyTimes (xs `snoc` a)
    SetManyTimes xs <> SetManyTimes ys = SetManyTimes (xs <> ys)

instance Monoid (SetOnce a) where
    mempty = NotSet
```



Rendering/erroring can then happen outside of the code.